### PR TITLE
Test suite for the gspread script

### PR DESCRIPTION
**Pull request by LogicAL007**
-Tests for the connection with the Weather API.
-Tests to check if the expected column names are present.
 -Tests to check the data types of columns
-Tests to check if data in certain columns falls within valid ranges. 
-Tests related to successful worksheet creation.

### CODE REVIEW

File: __pycache__/my_gspread.cpython-310.pyc
Comment (elseagle) [inline __pycache__/my_gspread.cpython-310.pyc:1]: Kindly delete/add binary files to .gitignore

File: README.md [hunk 1]
```diff
@@ -1,77 +1,31 @@
...
-   - Pandas library
-   - Dotenv
-
-##  Data Source
-The data used was the first 500 rows of a dataset obtained from Kaggle. click on this [link](https://www.kaggle.com/datasets/lissetteg/ecommerce-dataset) to access the file.
+This repository contains a pytest test suite for testing the upload of a json file from weatherAPI to a Google Sheet using the gspread library.
 
 ## Prerequisites
 
-Before using this script, ensure you have the following prerequisites in place:
```
Comment (elseagle) [inline README.md:20]: Why did you delete your former readme description?

File: my_gspread.py
Comment (elseagle) [inline my_gspread.py:1]: When you run your test, does it not trigger this whole script?

Try to use 
```
if __main__ == "__name__"

File: my_gspread.py [hunk 2]
```diff
@@ -1,23 +1,59 @@
...
 import os 
 import logging
+import requests
+import json
+import datetime as dt
 
+# access keys using dotenv
 load_dotenv()
-API_key =  os.getenv('Project_key')
+gspread_key =  os.getenv('Project_key')
```
Comment (elseagle) [inline my_gspread.py]: Use upper case for your environment variable names e.g PROJECT_KEY

File: my_gspread.py
Comment (elseagle) [inline my_gspread.py:1]: the name of this sheet is not good enough, my_spread is somewhat vague. If this is your main script or an util or helper? If it's your main or entry point, you can simply name this app or main.

File: test_my_gspread.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import os
+import requests
+import json
+import datetime as dt
+
+from my_gspread import get_or_create_worksheet
+from my_gspread import pass_json_into_df
+
+load_dotenv()
+gspread_key = os.getenv('Project_key')
```
Comment (elseagle) [inline test_my_gspread.py]: [Don't Repeat Yourself (DRY)](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)

This is something you did in the `my_gspread.py` already. Why not create a setup() or connect_sheet() that you can always call?

File: test_my_gspread.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+gc = gspread.service_account(gspread_key)
```
Comment (elseagle) [inline test_my_gspread.py]: I think I'd prefer, you don't do this. When writing tests, a good practice is to avoid anything that will call APIs or connect to an external network outside your environment. So the way to go is to use mocks, If you try to run a test that connects over an API you're not unit-testing your code because you're now testing the code of the API inside your own code, the goal of this project is unit testing not integration testing

File: test_my_gspread.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    assert worksheet.title == worksheet_name
+    assert spreadsheet.title == 'Gspread practice'
+    assert worksheet.row_count != 0
+
+def test_for_connection_with_API():
+    '''
+    This test for the successful connection with the weather API, by checking the response state code
+     '''
+    payload = {'Key': config['Key'], 'q': 'berlin', 'aqi': 'no'}
+    r = requests.get("http://api.weatherapi.com/v1/current.json", params=payload)
```
Comment (elseagle) [inline test_my_gspread.py]: Don't do this for unit tests, you're supposed to test units of your code

**Review by elseagle — CHANGES_REQUESTED** [review]
Solid work, we're getting closer to closing this! Kindly check out the new comments

File: README.md (hunk 1 above)
Comment (LogicAL007) [reply README.md:20]: oh i didnt delete it, it has already been merged with the main branch after you approved it, this is probably because i edited a copy of it for this new readme sir

File: my_gspread.py
Comment (LogicAL007) [reply my_gspread.py:1]: okay sir

File: my_gspread.py (hunk 2 above)
Comment (LogicAL007) [reply my_gspread.py]: I will change that now sir

**Review by elseagle — APPROVED** [review]

**Comment by elseagle** [comment]
You have merge conflicts to resolve, basically because you might have altered some files you worked on while this PR was open in another commit or branch and merged with the master/main branch. You need to update the two files affected to the version you need them to be in. After you've resolved it kindly merge.
